src/coreclr/jit/importer.cpp

```diff
@@ -1677,8 +1677,7 @@ GenTree* Compiler::impRuntimeLookupToTree(CORINFO_RESOLVED_TOKEN* pResolvedToken
         return slotPtrTree;
     }
 
-    slotPtrTree = gtNewIndir(TYP_I_IMPL, slotPtrTree, GTF_IND_NONFAULTING);
-    slotPtrTree->gtFlags &= ~GTF_GLOB_REF; // TODO-Bug?: this is a quirk. Can we mark this indirection invariant?
+    slotPtrTree = gtNewIndir(TYP_I_IMPL, slotPtrTree, GTF_IND_NONFAULTING | GTF_IND_INVARIANT);
```

Comment:
I don't see why it can't be invariant (and removed the GTF_GLOB_REF quirk). However, I most likely will have to unmark it as invariant in my "inlining for shared generics" work for certain cases, but it's irrelevant here.

Reply:
I would double-check this with somebody on the runtime team; getting this wrong can have subtle consequences.

Reply:
So from my understanding there are two types of runtime lookups:
1) Caller-side 
2) Callee-side

The caller side is typically not invariant (at least in JIT) and requires a helper call + expansion machinery (checking for init/dictionary resize).
The callee side is always invariant and is just a bunch of invariant loads on top of TypeCtx

It might be a bit more complicated if we implement inlining for shared generics where we're going to have both caller and callee lookups in the same method, but I will keep that in mind once I do that sort of inlining (as part of net9).

cc @jkotas to confirm.

Reply:
The generic dictionary slots that do not require helper-call/null check/size check can be treated as invariant. This change looks fine to me.

> So from my understanding there are two types of runtime lookups:
> Caller-side
> Callee-side

I am not sure what you mean. There is no distinction like this that I am aware of.

Reply:
> I am not sure what you mean. There is no distinction like this that I am aware of.

Yes, please, ignore that part. There are just 4 types of lookups:

1) Helper call + nullcheck + IND (fast path)
2) Helper call + nullcheck + sizecheck + IND (fast path)
3) nullcheck + IND (fast path)
4) IND (fast path)

I presume the helper call without the size check is invariant too (before we expand it).

Reply:
> nullcheck + INDs (fast path)

Nit: These do not exist. The null check needs a helper call slow path.